### ISSUE TITLE
[build] Ensure build is reproducible

### ISSUE THREAD

**Issue by aalmiray** [jreleaser/jreleaser#1024]
Requires following recommended practices for reproducible builds with Gradle, in particular:
- archives must have a deterministic entry order
- archive entries must use a well known timestamp (suggestion: use latest commit timestamp)

Requires upgrading Kordamp plugins and Gradle 7.

cc: @hboutemy

**Comment by aalmiray** [comment]
@hboutemy made some progress. What's missing:
- `JReleaserVersion` still uses `now()` instead of commit timestamp.
- releasable JARs have the following manifest entries
```
Manifest-Version: 1.0
Main-Class: org.jreleaser.cli.Main
Created-By: Gradle 6.9
Build-Date: 2022-11-16
Build-Time: 16:01:05+02:00
Build-Revision: 9a300df4ea057ca4d57d09714d97cfa9eb5e2c74
Build-Jdk-Spec: 11
Specification-Title: jreleaser
Specification-Version: 1.4.0-SNAPSHOT
Specification-Vendor: Kordamp
Implementation-Title: jreleaser
Implementation-Version: 1.4.0-SNAPSHOT
Implementation-Vendor: Kordamp
```
where `Build-Date:` and `Build-Time:` refer to `now()` instead of the commit timestamp.

**Comment by aalmiray** [comment]
`Build-Jdk-Spec:` is added automatically and can't be skipped at the moment. 
A new version of the Kordamp Gradle Plugins must be released to enable the following:
- an option to skip `Build-Jdk-Spec`
- use commit timestamp instead of `now()` for `Build-Date:` and `Build-Time:`

A new release of Kordamp Gradle Plugins requires:
- updating the build to latest Gradle 7
- updating dependencies (latest jgit and grgit require Java 11 as a base)

These changes would break binary compatibility for `jdks-gradle-release` and `jreleaser-gradle-release` as they would require Gradle 7 and Java 11 to work. In which case these changes should be posted in a `2.x` release to ensure the whole build adheres to semver.

cc: @sormuras

**Comment by aalmiray** [comment]
Alright, managed to shave that yak and now JReleaser creates reproducible artifacts in its build, as far as I can tell.

```
Manifest-Version: 1.0
Main-Class: org.jreleaser.cli.Main
Created-By: Gradle 7.5.1
Build-Date: 2022-11-16
Build-Time: 16:29:44+02:00
Build-Revision: b72fdec3fa14a6a295555325f56590dbe6192c4f
Build-Jdk-Spec: 11
Specification-Title: jreleaser
Specification-Version: 1.4.0-SNAPSHOT
Specification-Vendor: Kordamp
Implementation-Title: jreleaser
Implementation-Version: 1.4.0-SNAPSHOT
Implementation-Vendor: Kordamp
```

The date and time now match the commit time for b72fdec. This however still bumps `jdks-gradle-plugin` and `jreleaser-gradle-plugin` to Gradle 7 and JDK 11 as a minimum. In strict sense posting a `1.4.0` release with these changes constitutes a compatibility break. However, Gradle consumers are usually faster in upgrading to latest Gradle (that may take care of Gradle 7.x). The question remains with JDK 11.

We could play it extremely safe and delay this upgrade til JReleaser `2.x` (for which we have no date nor plan yet), or bump the Gradle plugins (YOLO!) while keeping the rest of the build compatible with JDK 8 as it currently is.

**Comment by hboutemy** [comment]
IIUC, you now have checked that 2 builds done on the same machine give you the same output, by comparing binary output: that's it?
if that step is done, I see 2 possible next steps:
1. stay simple, release early, release often: do a release and see if a rebuild done on https://github.com/jvm-repo-rebuild/reproducible-central brings the same output
2. try another stronger personal test before releasing: try to compare 2 build outputs done in seriously different build conditions: either 2 different machines (like local and CI, or another personal machine), or use a Docker container to do the other build

**Comment by aalmiray** [comment]
Well, I'm almost there but can't make a release just yet because changes bring a binary incompatibility to the 2 Gradle plugins we currently release as part of the release train.

We're not yet ready to bump major to 2. I'll see if I can find a way to have reproducible builds while keeping BC with Gradle 6 and Java 8.

**Comment by aalmiray** [comment]
Well, I'm almost there but can't make a release just yet because changes bring a binary incompatibility to the 2 Gradle plugins we currently release as part of the release train.

We're not yet ready to bump major to 2. I'll see if I can find a way to have reproducible builds while keeping BC with Gradle 6 and Java 8.

**Comment by aalmiray** [comment]
Released in v1.4.0 -> https://github.com/jreleaser/jreleaser/releases/tag/v1.4.0
